### DISCUSSION TITLE
Output h264 directly to disk

### DISCUSSION THREAD

**Discussion by dchoma-sy** [bluenviron/gortsplib#105] [Questions]
Forgive my huge gaps in knowledge. But I was wondering if it's possible to take an rtph264 track and write the NALUs directly to disk.

I've been playing with the [client-read-h264-save-to-disk](https://github.com/aler9/gortsplib/tree/main/examples/client-read-h264-save-to-disk) example. But I seem to be getting something wrong (lots of smearing and artifacts), and I suspect it's an issue with the TS encoder.

In the examples, the process seems to be a matter of using the client to decode the NALUs from the RTP packets, and then using another package to encode them to other formats (mpegts, images, etc). Is it possible to do this without encoding into other formats?

Thanks for a great library!

**Comment by aler9** [comment]
Hello,

To eliminate smearing and artifacts, you have to set `Client.Transport` to `TransportTCP` or increase `ReadBufferCount`.

Raw H264 NALUs don't contain timestamps. They can be decoded but they can't be played in the correct order and with the correct timing. MPEG-TS adds the timestamp information that allows to play them correctly.

**Reply by dchoma-sy** [reply]
Thanks for the quick reply @aler9!

I tried both increasing the `ReadBufferCount` and using the `TransportTCP` transport. Neither made an impact on the output file. The artifacts seem to occur consistently, starting about 15 seconds into the TS file.

Just to confirm that my source stream isn’t the issue, I tested a similar process (rtsp -> TS files) with ffmpeg. The output files look fine.

I’m going to try a few other examples in your repo. But if you have any other wisdom to share, I’d love to try some more tests.

Thanks again!
